rosbag2_compression/src/rosbag2_compression/sequential_compression_writer.cpp

```diff
@@ -74,8 +75,54 @@ SequentialCompressionWriter::~SequentialCompressionWriter()
   reset();
 }
 
+void SequentialCompressionWriter::compression_thread_fn()
+{
+  // Every thread needs to have its own compression context for thread safety.
+  auto compressor = compression_factory_->create_compressor(
+    compression_options_.compression_format);
+
+  std::mutex mutex;
+  std::unique_lock<std::mutex> lock(mutex);
+
+  if (!compressor) {
+    throw std::runtime_error{
+            "Cannot compress message; Writer is not open!"};
+  }
+
+  while (compression_is_running_) {
+    std::shared_ptr<rosbag2_storage::SerializedBagMessage> message;
+    std::string file;
+    compressor_condition_.wait(lock);
+    {
+      std::lock_guard<std::mutex> queue_lock(compressor_mutex_);
+      if (!compressor_message_queue_.empty()) {
+        message = compressor_message_queue_.front();
+        compressor_message_queue_.pop();
+      } else if (!compressor_file_queue_.empty()) {
+        file = compressor_file_queue_.front();
+        compressor_file_queue_.pop();
+      }
+    }
+
+    if (message) {
+      compress_message(*compressor, message);
+
+      {
+        // Now that the message is compressed, it can be written to file using the
+        // normal method.
+        std::lock_guard<std::recursive_mutex> storage_lock(
+          storage_mutex_);
+        SequentialWriter::write(message);
```

Comment:
Sorry for suddenly jumping in here (as I am very new to project, and still getting familiar with code base):
As far as i understand, by introducing multiple threads concurrently racing for messages, would change messages sequential order in the bag.
I would even say, it became non-deterministic.
How important is it?

(please feel free to ignore me)

Reply:
wow, thanks for the clarification. if messages are sorted by timestamp at replay, it is not very important indeed.

Reply:
It is true that the order that messages are received is not necessarily the order in which they were sent. The behavior now is to replay them according to their timestamp - note that this timestamp is created at the time of receipt. So, by design rosbag2 will playback messages in the order they were received.

>  I'd say it's still deterministic

This isn't true. Deterministic in this point would mean that the same inputs will always cause the same order of writing to disk. Depending on how the compression threads are scheduled and given priority by the kernel, they may complete in different times and therefore have a different write queue order on subsequent runs.

It's not important though, given that playback is sorted by timestamp - just wanted to make the distinction.

> I don't think there's a need to worry about storage backends that don't exist

I should also note that we are evaluating another storage backend now https://github.com/ros2/rosbag2/pull/556. This shouldn't block us from moving forward, but it's something to note.

Reply:
> It's true that this will mean the order the messages are received is not necessarily the order in which they will be written to the file. I'd say it's still deterministic, it's just not something that a human would be able to realistically determine. wink
> 
> With the current default SQLite-based storage, this doesn't actually matter. When messages are read from the bag files, they are ordered based on message timestamp (https://github.com/ros2/rosbag2/blob/master/rosbag2_storage_default_plugins/src/rosbag2_storage_default_plugins/sqlite/sqlite_storage.cpp#L367), which is independent of the order they were written to the file. As an interesting side note, due to factors like network latency, there is no guarantee that messages from multiple different nodes will be delivered to rosbag2 in the order that they were created, so the playback order could be different from the order they were received even if everything is single-threaded.
> 
> Of course, this isn't necessarily true of any other storage backends, but I don't think there's a need to worry about storage backends that don't exist. ROS1's bag format notably played back messages in the order they were written in the file, which, in my experience, was mostly undesirable -- in practice, I almost always needed to run a script on my bag files to rewrite them so that they messages were in timestamp order.

@pjreed A follow-up to the order of messages. In your description of ROS1, are you referring to the Python or C++ implementation?

We looked into ROS1 Python [rosbag.bag](https://github.com/ros/ros_comm/blob/noetic-devel/tools/rosbag/src/rosbag/bag.py) and its C++ counterpart rosbag_storage [view](https://github.com/ros/ros_comm/blob/noetic-devel/tools/rosbag_storage/src/view.cpp) and came to the conclusion that both will merge messages from different chunks sorted by index time, but only the C++ code does sort messages within a chunk by index time, while the Python code did not:

- https://github.com/ros/ros_comm/blob/noetic-devel/tools/rosbag/src/player.cpp#L184
- https://github.com/ros/ros_comm/blob/noetic-devel/tools/rosbag/src/player.cpp#L315
- https://github.com/ros/ros_comm/blob/noetic-devel/tools/rosbag_storage/src/view.cpp#L124
- https://github.com/ros/ros_comm/blob/noetic-devel/tools/rosbag_storage/src/view.cpp#L320
- https://github.com/ros/ros_comm/blob/noetic-devel/tools/rosbag_storage/src/view.cpp#L301

We created an MR to fix the Python code. It would be great, if you could take a look and verify our findings.

https://github.com/ros/ros_comm/pull/2089

Reply:
Right, I was specifically thinking of the Python implementation, since that's what gets used when you run `rosbag play`. I reviewed that PR and that looks like a really good change to me, thanks!